transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java

```diff
@@ -0,0 +1,283 @@
+            if (localReadAmount > 0) {
+                byteBuf.writerIndex(byteBuf.writerIndex() + localReadAmount);
+            }
+
+            final IOUringRecvByteAllocatorHandle allocHandle =
+                    (IOUringRecvByteAllocatorHandle) event.getAbstractIOUringChannel().unsafe()
+                                                          .recvBufAllocHandle();
+            final ChannelPipeline pipeline = event.getAbstractIOUringChannel().pipeline();
+
+            allocHandle.lastBytesRead(localReadAmount);
+            if (allocHandle.lastBytesRead() <= 0) {
+                // nothing was read, release the buffer.
+                byteBuf.release();
+                byteBuf = null;
+                break;
+            }
+
+            allocHandle.incMessagesRead(1);
+            //readPending = false;
+            pipeline.fireChannelRead(byteBuf);
+            byteBuf = null;
+            allocHandle.readComplete();
+            pipeline.fireChannelReadComplete();
+            event.getAbstractIOUringChannel().executeReadEvent();
+        break;
+        case WRITE:
+            System.out.println("EventLoop Write Res: " + res);
+            System.out.println("EventLoop Fd: " + event.getAbstractIOUringChannel().getSocket().intValue());
+            System.out.println("EventLoop Pipeline: " + event.getAbstractIOUringChannel().eventLoop());
+            ChannelOutboundBuffer channelOutboundBuffer = event
+                    .getAbstractIOUringChannel().unsafe().outboundBuffer();
+            //remove bytes
+            int localFlushAmount = res;
+            if (localFlushAmount > 0) {
+                channelOutboundBuffer.removeBytes(localFlushAmount);
+            }
+
+            try {
+                event.getAbstractIOUringChannel().doWrite(channelOutboundBuffer);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        break;
+        case TIMEOUT:
+            if (res == ETIME) {
+                prevDeadlineNanos = NONE;
+            }
+
+        break;
+        case POLL_EVENTFD:
+            pendingWakeup = false;
+            //Todo eventId is already used
+            long eventId = incrementEventIdCounter();
+            event.setId(eventId);
+            event.setOp(EventType.POLL_EVENTFD);
+            addNewEvent(event);
+            submissionQueue.addPoll(eventId, eventfd.intValue(), event.getOp());
+        case POLL_LINK:
+            //Todo error handling error
+            System.out.println("POLL_LINK Res: " + res);
+        break;
+        }
+        this.events.remove(event.getId());
+    }
+
+    public RingBuffer getRingBuffer() {
+        return ringBuffer;
+    }
+
+    @Override
+    protected void wakeup(boolean inEventLoop) {
+        if (!inEventLoop && nextWakeupNanos.getAndSet(AWAKE) != AWAKE) {
+            // write to the evfd which will then wake-up epoll_wait(...)
+            Native.eventFdWrite(eventfd.intValue(), 1L);
+        }
```

Comment:
Looks like comments about the wake up and blocking behaviour is a bit all over the place, so I'll add more… here.

You created the `Native.ioUringRegisterEventFd` plumbing, but you are not using it anywhere.
As I read the `io_uring_register` documentation, the registered eventfd will be notified when there are completions ready to be processed.

Given this, the problem of having two separate things to block on for one thread, could then be solved by having the task queue also implement its blocking behaviour using eventfd, and then the io_uring and the task queue could both use the same eventfd.
The single event loop thread would never block on either directly - only on the eventfd.

I think this would give us the blocking behaviour we want, and we would side-step the problem of getting CQEs from eventfd reads inside io_uring, there would be no need for enqueueing extra SQEs for the purpose of unblocking the io_uring - no timeouts, not eventfd reads - plus we'd avoid the bugs you found in that part of the kernel code (assuming I've correctly understood the nature of those bugs).

Reply:
@chrisvest I'm not sure I'm understanding this right, I'm guessing the tasks queue execute eventfd.write when there is new task added, right? just like in the wakupe method right?
how do we deal with schedule tasks without io_uring timeouts?

Reply:
@chrisvest it's an interesting idea but if I'm wondering the same as @1Jo1 re scheduled wakeups. We would need to put the eventfd in an epoll along with a timerfd I think (as is done currently for epoll transport). I also think that this would entail a greater number of syscalls which is the primary thing we're trying to avoid. IIUC we would still need to make io_uring_enter calls to submit events, and would lose the advantage of combining this with the blocking.

Relative to that, the cost/inconvenience of ignoring the occasional otherwise-redundant CQE seems insignificant. And as mentioned in other comments I think there's a way we can use io_uring timeout (with 0) for the wakeup without requiring expensive sync of the SQ. Then eventfd goes away altogether.

Reply:
> I'm not sure I'm understanding this right, I'm guessing the tasks queue executes eventfd.write when there is a new task added, right? just like in the wakeup method?

Yes, that's what I had in mind.

> how do we deal with scheduled tasks without io_uring timeouts?

Ah, right, timeouts make sense for scheduled tasks.

> And as mentioned in other comments I think there's a way we can use io_uring timeout (with 0) for the wakeup without requiring expensive sync of the SQ. Then eventfd goes away altogether.

I've read those comments, but I still don't understand how it's supposed to work. Does it include a way for the submission of tasks to unblock the event loop thread (how?), or will the thread poll the task queue periodically via timeout wakeups? Or is it spinning?

Reply:
@chrisvest the overall structure would remain similar to `EpollEventLoop`'s. Before the event loop enters a blocking `io_uring_enter` call, it sets the `nextWakeupNanos` atomic long to `AWAKE` before which task-submission wakeups can't occur (since they are blocked by a CAS of that value). At this point, whoever wins the CAS of `nextWakeupNanos` back to `AWAKE` "owns" the next position in the SQ.

Instead of calling `eventfd_write`, the `wakeup` method will add a zero-timeout entry to the SQ (I propose having a dedicated "constant" SQE in the SQ array for this), increment the queue's head by 1 and call `io_uring_enter` with `to_submit=1` which will wake up the event loop currently blocked in its own `io_uring_enter`. Upon resuming, the EL will know whether it was woken by task submission thread based on whether the zero-timeout CQE is present. If not, it will also detect the race based on whether it wins its immediate CAS of `nextWakeupNanos` back to `AWAKE`.

In the first case it just skips over the next position in the SQ. In the second (race) case, some additional care is needed if the zero-timeout CQE has still not been seen at the point that the EL wants to submit new SQEs (i.e. the next time it calls io_uring_enter). I can think of some variations of how to handle this and not sure which is optimal overall performance-wise. One possibility is to have the waker thread's increment of the SQ head ptr be atomic, have the EL track the fact it's in this state (with simple flag) and for it to also perform atomic increments of the SQ head itself while it is. To make this safe the EL would actually first (over)write the same zero-timeout SQE index to the SQ to ensure that it's set. It shouldn't matter which order the kernel receives the corresponding submission calls, their `to_submit` values will add up to the correct value. WDYT?

Reply:
@njhill I think this makes sense. Thanks for writing it up. I don't see anything obviously wrong with the waker atomically bumping the pointer, but I also haven't investigated it sufficiently to be confident that it's correct. Lock-free algorithms are extremely subtle. Overall I agree with this approach, and think it would be the best solution if we can make it work.

Reply:
@njhill timeout wakeup lock free sounds interesting

>  Before the event loop enters a blocking io_uring_enter call, it sets the nextWakeupNanos atomic long to AWAKE

I guess you mean to set atomic long to NONE or scheduledTask nanoseconds right?  maybe I missed something